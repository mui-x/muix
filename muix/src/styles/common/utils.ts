export function deepMerge(target, ...sources) {
  if (!sources.length) return target

  const source = sources.shift()
  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!target[key]) target[key] = {}
        deepMerge(target[key], source[key])
      } else 
        target[key] = source[key]
    }
  } else
    throw 'Cannot merge object and non object props'

  return deepMerge(target, ...sources);
}

const isObject = item => item && typeof item === 'object' && !Array.isArray(item)
