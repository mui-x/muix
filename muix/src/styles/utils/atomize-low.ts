import { wrapPseudoPrefixes } from './wrap-pseudo-prefixes'
import { TEngine } from 'reactxx-typings'

// platform dependent 
import { platform } from 'reactxx-styles'

//*******************************************************
//        adjustAtomizedLow

// linearize and atomize "ruleset"
// result is added to 'atomizedVariants'
export const adjustAtomizedLow = (
    ruleset: TEngine.Rulesets,
    atomizedVariants: TEngine.QueryableItems, path: string,
    pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    processTree(ruleset, atomizedVariants, path, pseudoPrefixes, conditions)
}

export function isToAtomizeArray(obj): obj is TEngine.Ruleset[] { return obj && !(obj as TEngine.Queryables).$r$ && Array.isArray(obj) }
export function isAtomized(obj): obj is TEngine.Queryables { return !obj || (obj as TEngine.Queryables).$r$ }
export function isDeferred(obj): obj is TEngine.Deferred { return obj && (obj as TEngine.Deferred).$d$ }
export function isTemporary(obj): obj is TEngine.TempProc { return obj && (obj as TEngine.TempProc).$t$ }
export function isToAtomize(obj): obj is TEngine.ToAtomize { return obj && !(obj as TEngine.Queryables).$r$ && !(obj as TEngine.Deferred).$d$ && !(obj as TEngine.TempProc).$t$ }

//*******************************************************
//        makeTemporary
export const makeTemporary = (proc: TEngine.TempProc) => {
    proc.$t$ = true
    return proc as any
}

//*******************************************************
//        processTree
export const processTree = (
    value: TEngine.Rulesets, atomizedVariants: TEngine.QueryableItems,
    path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {

    const indexToPath = (idx: number) => items.length <= 1 ? '' : `[${idx}]`

    const items = isToAtomizeArray(value) ? value : [value]
    items.forEach((it, idx) => {
        if (!it) return
        if (isTemporary(it)) {
            it(atomizedVariants, path + indexToPath(idx), pseudoPrefixes, conditions)
        } else if (isDeferred(it)) {
            atomizedVariants.push(it)
        } else if (isAtomized(it)) {
            if (pseudoPrefixes && pseudoPrefixes.length > 0)
                throw 'Web pseudo properties cannot contain already atomized value'
            for (const itt of it) {
                if (isDeferred(itt))
                    throw 'isDeferred(itt)'
                if (!conditions || conditions.length === 0) //... and no conditions => done
                    pushToAtomizedVariants(atomizedVariants, itt, null)
                else { // ... and conditions => merge conditions
                    const ittCopy = Array.isArray(itt) ? [...itt] : { ...itt }
                    pushToAtomizedVariants(atomizedVariants, ittCopy,
                        itt.conditions ? [...conditions, ...itt.conditions] : conditions)
                }
            }
        } else { // it: toAtomize
            // if (window.__TRACE__)
            //     it = deepClone(it)
            processPseudosAndAtomize(
                it,
                atomizedVariants, path + indexToPath(idx),
                pseudoPrefixes, conditions
            )
        }
    })
}

//*********************************************************
//  PRIVATE
//*********************************************************

//*******************************************************
//        processAllInnerVariants

// recursivelly process web pseudos (:hover, :active) and its queryable parts

const processPseudosAndAtomize = (
    ruleset: TEngine.ToAtomize,
    atomizedVariants: TEngine.QueryableItems, path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions,
) => {

    const inner: TEngine.QueryableItems = []

    const toAtomizePure = {}
    let empty = true
    // process and remove conditional parts of self and web pseudo (e.g. :hover etc.)
    // push them to 'inner'
    for (const p in ruleset) {
        if (p.charAt(0) === '$') continue
        const value = ruleset[p] as TEngine.Rulesets
        if (typeof value === 'object' || isTemporary(value)) { // temporary, deffered, toAtomize, atomized
            processTree(value, inner, `${path}/${p}`, [...pseudoPrefixes, p], conditions)
        } else {
            empty = false
            toAtomizePure[p] = value
        }
    }

    // atomize pure ruleset tree with pseudo (and without temporary, deffer's etc.),
    // then push to result
    if (!empty)
        atomizePure(atomizedVariants, wrapPseudoPrefixes(toAtomizePure, pseudoPrefixes), conditions, path)

    // push 'inner' to result AFTER toAtomizePure
    if (inner.length > 0)
        Array.prototype.push.apply(atomizedVariants, inner)

}

// *********** utils

const pushToAtomizedVariants = (
    atomizedVariants: TEngine.QueryableItems,
    variant: TEngine.Queryable, conditions: TEngine.Conditions
) => {
    if (!variant) return
    if (conditions && conditions.length > 0) variant.conditions = conditions
    atomizedVariants.push(variant)
}

const atomizePure = (
    atomizedVariants: TEngine.QueryableItems, pureRuleset: TEngine.ToAtomize,
    conditions: TEngine.Conditions, path: string
) => {
    if (!pureRuleset) return
    if (isDeferred(pureRuleset)) {
        throw 'something wrong'
        atomizedVariants.push(pureRuleset)
        return
    }
    // platform specific atomize: e.g. for WEB convert ruleset to list of atomized CSS classes
    const variant = platform.toPlatformAtomizeRuleset(pureRuleset, path)
    pushToAtomizedVariants(atomizedVariants, variant, conditions)
}

export function deepClone(obj) {
    if (
        obj === undefined || obj === null || typeof obj === 'string' || typeof obj === 'undefined' ||
        typeof obj === 'function' || typeof obj === 'number' || typeof obj === 'boolean'
    ) return obj

    if (Array.isArray(obj)) {
        const res = obj.map(o => deepClone(o))
        if (obj['$r$']) res['$r$'] = true
        return res
    }

    if (typeof obj === 'object') {
        const res = {}
        for (const p in obj) res[p] = deepClone(obj[p])
        return res
    }

    throw 'something wrong'
}