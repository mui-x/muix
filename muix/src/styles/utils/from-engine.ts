import { TEngine, TTyped } from 'reactxx-typings';

export const TAsEngineClassName = <R extends TTyped.Shape>(r: TEngine.RulesetOrCreator) => r as TTyped.ClassNameSimple<R>
export const TAsTypedClassName = (r: TTyped.RulesetOrCreator) => r as TEngine.RulesetOrCreator

export const TAsEngineSheet = <R extends TTyped.Shape>(r: TEngine.SheetOrCreator) => r as any as TTyped.SheetSimple<R>
export const TAsTypedSheet = (r: TTyped.PartialSheetOrCreator) => r as any as TEngine.Sheet

export const TAsEngineStyle = <R extends TTyped.Shape>(r: TEngine.StyleOrCreator) => r as any as TTyped.StyleSimple<R>
export const TAsTypedStyle = (r: TTyped.StyleOrCreator) => r as any as TEngine.Style

// export function isReactXXComponent(obj): obj is TEngine.IsReactXXComponent {
//     return (obj as TEngine.IsReactXXComponent).$c$
// }