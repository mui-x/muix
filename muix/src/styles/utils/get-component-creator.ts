//import React from 'react'
import { TTyped, TComponents } from 'reactxx-typings';

export const getComponentCreator = <R extends TTyped.Shape>(
    authorDisplayName: string, authorConfig: TComponents.AuthorConfig<R>, getComp: TComponents.GetComponent<R>, par?
) => (userDisplayName?: string, userConfig?: TComponents.UserConfig<R>) => {
    const displayName = userDisplayName || authorDisplayName
    const Comp = getComp(authorConfig, displayName, userConfig, par)
    Comp.displayName = displayName
    //Comp.$c$ = true
    return Comp
}

export const getComponentCreatorUntyped = (authorDisplayName, authorConfig, getComp, par?) =>
    getComponentCreator(authorDisplayName, authorConfig, getComp, par) as any