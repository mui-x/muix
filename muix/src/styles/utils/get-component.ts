import warning from 'warning'

import { TTyped, TComponents } from 'reactxx-typings';

export const getComponent = <R extends TTyped.Shape>(
    getComp: TComponents.GetComponent<R>, configss?: TComponents.ComponentConfigss<R>
) => {
    const config = mergeConfig(configss)
    config.getComponent = getComp

    const Comp = config.Comp = getComp(config)
    Comp.displayName = config.displayName
    Comp.config = config

    return Comp
}

export const getComponentUntyped = getComponent as any

const mergeConfig = <R extends TTyped.Shape>(inputss: TComponents.ComponentConfigss<R>, config?: TComponents.Config<R>) => {
    if (!config) config = {}
    if (!inputss) return config
    const doMerge = (input: TComponents.ComponentConfig) => {
        if (!input) return
        for (const p in input) {
            if (p.charAt(0) !== '$')
                config[p] = input[p]
            else {
                if (!config[p]) config[p] = []
                config[p].push(input[p])
            }
        }
    }
    if (Array.isArray(inputss))
        for (const inputs of inputss)
            if (Array.isArray(inputs))
                for (const input of inputs)
                    doMerge(input)
            else
                doMerge(inputs)
    else
        doMerge(inputss)
    return config
}
