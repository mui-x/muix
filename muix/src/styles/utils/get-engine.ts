import { TTyped } from 'reactxx-typings';
import { 
    THEMED, QUERY, COMPILE, STYLE, WIDTH, WEB, NATIVE, HOT, IF, IFELSE, COMPONENT
} from '../queryable/index'

export const getTypedEngine = <R extends TTyped.Shape>() => engine as TTyped.TypedEngine<R>

const engine = {
    WEB,
    NATIVE,
    STYLE,
    COMPILE,
    HOT,
    IF,
    IFELSE,
    WIDTH,
    THEMED,
    QUERY,
    COMPONENT,
}

