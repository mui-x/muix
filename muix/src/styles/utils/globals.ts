import { TExtensions, TEngine, TTyped } from 'reactxx-typings';
import { WidthStore } from '../queryable/$widths/store'

export interface Platform extends TExtensions.Platform {
    toPlatformAtomizeRuleset?: TEngine.ToPlatformAtomizeRuleset
    applyLastwinsStrategy?: TEngine.ApplyLastwinsStrategy
    finalizeClassName?: (values: TEngine.AtomicLow) => TEngine.AtomicFinal
    getDefaultTheme?: () => any
    styleProps?: TTyped.StyleProps
}

export const resetPlatform = () => {
    for (const p in platform) delete platform[p]
}

export const assignPlatform = (gl: Platform) => Object.assign(platform, gl)

export const platform: Platform = {}

export const initGlobals = (force: boolean, platformDependentInit: () => void) => {
    if (force) resetPlatform()
    if (platform._styles) return

    platform._styles = {
        $cache: {},
        defaultTheme: platform.getDefaultTheme && platform.getDefaultTheme(),
        instanceIdCounter: 0,
        configIdCounter: 0,
    }
    // fill e.g. actWidth
    platformDependentInit()
    // init store with act width
    platform._styles.widthsStore = new WidthStore()
}