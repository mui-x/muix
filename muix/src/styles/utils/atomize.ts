import warning from 'warning'

import { TEngine } from 'reactxx-typings'

import { adjustAtomizedLow, isDeferred, deepClone } from './atomize-low'
import { createWithTheme } from './create-with-theme'
import { isStyledComponentTemp } from '../hooks/use-styled-components'

// mutable
export const atomizeSheet = (sheet: TEngine.SheetOrCreator, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    if (window.__TRACE__)
        sheet = deepClone(sheet)
    const sh: TEngine.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const value = sh[p]
        const subPath = (path ? path + '.' : '') + p
        if (isStyledComponentTemp(value)) {
            warning(p.charAt(0)===p.charAt(0).toUpperCase(), 'Component definition name must start with uppercase letter')
            sh[p] = value(sh, p, subPath)
        } else {
            const at = atomizeRuleset(sh[p], theme, subPath)
            if (at) sh[p] = at
            else delete sh[p]
        }
    }
    return sh
}

// mutable
export const atomizeRuleset = (
    ruleset: TEngine.RulesetOrCreator, theme?, path: string = ''
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme)

    if (!rs) return null

    const list = wrapQueryables([])
    adjustAtomizedLow(rs, list, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapAtomicLow = (atomicLow: TEngine.AtomicLow) => {
    const res = [
        atomicLow
    ] as TEngine.Queryables
    res.$r$ = true
    return res
    // (queryables as TEngine.Queryables).$r$ = true
    // if (window.__TRACE__)
    //     queryables['toJSON'] = toJSON.bind(queryables)
    // return queryables as TEngine.Queryables
}


export const wrapQueryables = (queryables: TEngine.QueryableItems) => {
    (queryables as TEngine.Queryables).$r$ = true
    if (window.__TRACE__)
        queryables['toJSON'] = toJSON.bind(queryables)
    return queryables as TEngine.Queryables
}
function toJSON() {
    return (this as TEngine.Queryables).map(v => {
        if (isDeferred(v)) return 'DEFFERED'
        return v
    })
}

// mutable
export const atomizeStyle = (style: TEngine.Style, theme?) => {
    if (!style) return null

    style = createWithTheme(style, theme)

    const { $web, $native } = style

    if ($web || $native) style = { ...style }

    if (window.isWeb && $web)
        Object.assign(style, $web)
    else if (!window.isWeb && $native)
        Object.assign(style, $native)

    delete style.$web
    delete style.$native

    return style
}
