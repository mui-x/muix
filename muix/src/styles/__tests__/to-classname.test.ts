import { initPlatform, dump } from "./init-platform.t"
import { $W, $T, $V, $I, V, T, I, TTyped } from 'reactxx-typings'

import { getEngine } from '../utils/get-engine'

interface Shape {
  theme: { primary }
  sheetQuery: { opened: boolean }
}

const { THEMED, IF, IFELSE, WEB, NATIVE, STYLE, $toClassNames, ATOMIZE
} = getEngine<Shape>()

describe("TO CLASSNAMES", () => {

  type Par = { opened: boolean, theme: null }
  const sheetQuery = (opened: boolean) => ({ opened })
  const query = (opened: boolean, ...ruleset: TTyped.RulesetSimple[]) => dump($toClassNames<TTyped.RulesetIds>(sheetQuery(opened), ...ruleset))

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let ruleset

    it("01: null", () => dump($toClassNames(null, null)))
    it("02: null, {}, undefined", () => dump($toClassNames(null, undefined)))
    it("03: mixed", () => dump($toClassNames(null,
      ATOMIZE<T>({ color: 'red' }), // ruleset source
      ATOMIZE<T>({ color: 'blue' }, { color: 'yellow' }), // atomized ruleset
      IF<T>(p => true, { color: 'maroon' }), // temporary item
      $toClassNames<T>(null, ATOMIZE<T>({ color: 'green' })) // atomized ruleset
    )))
    it("04: query => opened", () => query(true,
      IF(p => p.opened, { color: 'red' }),
      IF(p => !p.opened, { color: 'blue' }),
    ))
    it("05: query => closed", () => query(false,
      IF(p => p.opened, { color: 'red' }),
      IF(p => !p.opened, { color: 'blue' }),
    ))
    it("06: query => closed => pre-atomized", () => query(false, ATOMIZE<T>(
      IF<T>(p => p.opened, { color: 'red' }),
      IF<T>(p => !p.opened, { color: 'blue' }),
    )))
    it("07: query => closed => pre-queried", () => query(false,
      $toClassNames<T>(sheetQuery(false),
        IF<T>(p => p.opened, { color: 'red' }),
        IF<T>(p => !p.opened, { color: 'blue' }),
      )
    ))
    it("08: static if and ifelse", () => {
      const inRender = (disabled: boolean) =>
        dump($toClassNames<T>(sheetQuery(false),
          IF<T>(disabled, { color: 'grey' }),
          IFELSE<T>(disabled, { backgroundColor: 'blue' }, { backgroundColor: 'green' }),
        ))
      inRender(true)
      inRender(false)
    })
  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
