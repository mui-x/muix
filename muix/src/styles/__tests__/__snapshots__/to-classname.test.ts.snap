// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TO CLASSNAMES ## NATIVE ## 01: null 1`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 01: null 2`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 01: null 3`] = `undefined`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 1`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 2`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 3`] = `undefined`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 1`] = `
Array [
  Object {
    "color": "red @$atomize",
  },
  Object {
    "color": "blue @$atomize[0]",
  },
  Object {
    "color": "yellow @$atomize[1]",
  },
  Object {
    "color": "maroon @/$if",
  },
  Object {
    "color": "green @$atomize",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 2`] = `
Object {
  "color": Object {
    "tracePath": "$atomize",
    "value": "green",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 3`] = `
Object {
  "color": "green",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 1`] = `
Array [
  Object {
    "color": "red @/$if",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 2`] = `
Object {
  "color": Object {
    "tracePath": "/$if",
    "value": "red",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 3`] = `
Object {
  "color": "red",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 1`] = `
Array [
  Object {
    "color": "blue @/$if",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 2`] = `
Object {
  "color": Object {
    "tracePath": "/$if",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 1`] = `
Array [
  Object {
    "color": "blue @$atomize[1]/$if",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 2`] = `
Object {
  "color": Object {
    "tracePath": "$atomize[1]/$if",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 1`] = `
Array [
  Object {
    "color": "blue @/$if",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 2`] = `
Object {
  "color": Object {
    "tracePath": "/$if",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 1`] = `
Array [
  Object {
    "color": "grey @/$sif",
  },
  Object {
    "backgroundColor": "blue @/$ifelse-true",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 2`] = `
Object {
  "backgroundColor": Object {
    "tracePath": "/$ifelse-true",
    "value": "blue",
  },
  "color": Object {
    "tracePath": "/$sif",
    "value": "grey",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 3`] = `
Object {
  "backgroundColor": "blue",
  "color": "grey",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 4`] = `
Array [
  Object {
    "backgroundColor": "green @/$ifelse-false",
  },
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 5`] = `
Object {
  "backgroundColor": Object {
    "tracePath": "/$ifelse-false",
    "value": "green",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 08: static if and ifelse 6`] = `
Object {
  "backgroundColor": "green",
}
`;

exports[`TO CLASSNAMES ## WEB ## 01: null 1`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 01: null 2`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 01: null 3`] = `undefined`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 1`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 2`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 3`] = `undefined`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 1`] = `
Array [
  Array [
    ".a { color:red } @$atomize",
  ],
  Array [
    ".b { color:blue } @$atomize[0]",
  ],
  Array [
    ".c { color:yellow } @$atomize[1]",
  ],
  Array [
    ".e { color:maroon } @/$if",
  ],
  Array [
    ".d { color:green } @$atomize",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 2`] = `
Array [
  ".d { color:green } @$atomize",
]
`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 3`] = `"d"`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 1`] = `
Array [
  Array [
    ".a { color:red } @/$if",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 2`] = `
Array [
  ".a { color:red } @/$if",
]
`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 3`] = `"a"`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 1`] = `
Array [
  Array [
    ".b { color:blue } @/$if",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 2`] = `
Array [
  ".b { color:blue } @/$if",
]
`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 3`] = `"b"`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 1`] = `
Array [
  Array [
    ".b { color:blue } @$atomize[1]/$if",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 2`] = `
Array [
  ".b { color:blue } @$atomize[1]/$if",
]
`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 3`] = `"b"`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 1`] = `
Array [
  Array [
    ".b { color:blue } @/$if",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 2`] = `
Array [
  ".b { color:blue } @/$if",
]
`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 3`] = `"b"`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 1`] = `
Array [
  Array [
    ".a { color:grey } @/$sif",
  ],
  Array [
    ".b { background-color:blue } @/$ifelse-true",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 2`] = `
Array [
  ".b { background-color:blue } @/$ifelse-true",
  ".a { color:grey } @/$sif",
]
`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 3`] = `"b a"`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 4`] = `
Array [
  Array [
    ".c { background-color:green } @/$ifelse-false",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 5`] = `
Array [
  ".c { background-color:green } @/$ifelse-false",
]
`;

exports[`TO CLASSNAMES ## WEB ## 08: static if and ifelse 6`] = `"c"`;
